R/gpt2.R

```diff
@@ -16,136 +16,138 @@
     att <- att$softmax(dim = -1)
     att <- self$attn_dropout(att)
+
     y <- att$matmul(v) # (B, nh, T, T) x (B, nh, T, hs) -> (B, nh, T, hs)
-    y <- y$transpose(1, 2)$contiguous()$view(c(B, T, C)) # re-assemble all head outputs side by side
+    y <- y$transpose(2, 3)$contiguous()$view(c(b, t, c)) # re-assemble all head outputs side by side
 
     # output projection
     y <- self$resid_dropout(self$c_proj(y))
     y
+  },
+  reset_parameters = function(initializer_range = 0.02) {
+    nn_init_normal_(self$c_attn$weight, mean = 0, std = initializer_range)
+    nn_init_zeros_(self$c_attn$bias)
+    nn_init_normal_(self$c_proj$weight, mean = 0, std = initializer_range/sqrt(2 * self$n_layer))
+    nn_init_zeros_(self$c_proj$bias)
   }
 )
 
 nn_gpt2_mlp <- nn_module(
-  initialize = function(n_embd, resid_pdrop, initializer_range) {
+  initialize = function(n_embd, pdrop) {
     self$c_fc <- nn_linear(n_embd, 4 * n_embd)
     self$c_proj <- nn_linear(4 * n_embd, n_embd)
-    self$act <- nn_new_gelu()
-    self$dropout <- nn_dropout(resid_pdrop)
+    self$act <- nn_gelu(approximate = "tanh")
+    self$dropout <- nn_dropout(pdrop)
 
-    nn_init_normal_(self$c_fc$weight, mean = 0, std = initializer_range)
-    nn_init_zeros_(self$c_fc$bias)
-    nn_init_normal_(self$c_proj$weight, mean = 0, std = initializer_range)
-    nn_init_zeros_(self$c_proj$bias)
+    self$reset_parameters()
   },
   forward = function(x) {
     x |>
       self$c_fc() |>
       self$act() |>
       self$c_proj() |>
       self$dropout()
+  },
+  reset_parameters = function(initializer_range = 0.02) {
+    nn_init_normal_(self$c_fc$weight, mean = 0, std = initializer_range)
+    nn_init_zeros_(self$c_fc$bias)
+    nn_init_normal_(self$c_proj$weight, mean = 0, std = initializer_range)
+    nn_init_zeros_(self$c_proj$bias)
   }
 )
 
 nn_gpt2_transformer_block <- nn_module(
-  initialize = function(n_embd, n_head, n_layer, n_positions, resid_pdrop, attn_pdrop,
-                        layer_norm_epsilon, initializer_range) {
-    self$ln_1 <- nn_layer_norm(n_embd, layer_norm_epsilon)
-    self$attn <- nn_gpt2_attention(n_embd, n_head, n_layer, n_positions, resid_pdrop, attn_pdrop,
-                                   initializer_range)
-    self$ln_2 <- nn_layer_norm(n_embd, layer_norm_epsilon)
-    self$mlp <- nn_gpt2_mlp(n_embd, resid_pdrop, initializer_range)
-
-    nn_init_zeros_(self$ln_1$bias)
-    nn_init_ones_(self$ln_1$weight)
-    nn_init_zeros_(self$ln_2$bias)
-    nn_init_ones_(self$ln_2$weight)
+  initialize = function(n_embd, n_head, n_layer, max_pos, pdrop) {
+    self$ln_1 <- nn_layer_norm(n_embd, eps = 1e-5)
+    self$attn <- nn_gpt2_attention(n_embd, n_head, n_layer, max_pos, pdrop)
+    self$ln_2 <- nn_layer_norm(n_embd, eps = 1e-5)
+    self$mlp <- nn_gpt2_mlp(n_embd, pdrop)
   },
   forward = function(x) {
-    x <- x + self$attn(self$ln_1(x)) + self$mlp(self$ln_2(x))
-    x
+    x <- x + self$attn(self$ln_1(x))
+    x + self$mlp(self$ln_2(x))
```

Comment:
These were not equivalent to the one liner: ```x + self$attn(self$ln_1(x)) + self$mlp(self$ln_2(x))``` because `ln_2` and `mlp` are applied to the modified `x`.